lnrpc/switchrpc/switch_server_test.go

```diff
@@ -393,6 +396,168 @@ func TestTrackOnion(t *testing.T) {
 	}
 }
 
+// TestBuildOnion is a unit test that verifies the behavior of the BuildOnion
+// RPC handler, ensuring that it correctly validates inputs, handles dependency
+// failures, and constructs a valid response upon success.
+func TestBuildOnion(t *testing.T) {
+	t.Parallel()
+
+	// Create a valid session key for the "provided key" test case.
+	validSessionKeyBytes := []byte{
+		1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
+		17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+	}
+
+	// Create a mock route to be returned by the processor.
+	// We need to generate some mock public keys for the hops.
+	privKey1, err := btcec.NewPrivateKey()
+	require.NoError(t, err)
+	pubKey1 := privKey1.PubKey()
+
+	privKey2, err := btcec.NewPrivateKey()
+	require.NoError(t, err)
+	pubKey2 := privKey2.PubKey()
+
+	// The mock route needs to have realistic values for amounts,
+	// timelocks, and channel IDs, otherwise sphinx packet construction
+	// will fail.
+	paymentAmt := lnwire.MilliSatoshi(10000)
+	finalCltv := uint32(40)
+
+	mockRoute := &route.Route{
+		TotalAmount: paymentAmt,
+		Hops: []*route.Hop{
+			{
+				PubKeyBytes:      route.NewVertex(pubKey1),
+				AmtToForward:     paymentAmt,
+				OutgoingTimeLock: finalCltv + 10,
+				ChannelID:        1,
+			},
+			{
+				PubKeyBytes:      route.NewVertex(pubKey2),
+				AmtToForward:     paymentAmt,
+				OutgoingTimeLock: finalCltv,
+				ChannelID:        2,
+			},
+		},
+	}
+
+	//nolint:ll
+	testCases := []struct {
+		name string
+
+		// setup is a function that modifies the server or request for a
+		// specific test case.
+		setup func(*testing.T, *mockRouteProcessor, *BuildOnionRequest)
+
+		// expectedErrCode is the gRPC error code we expect.
+		expectedErrCode codes.Code
+
+		// checkResponse is a function that validates the response on
+		// success.
+		checkResponse func(*testing.T, *BuildOnionResponse)
+	}{
+		{
+			name: "success new session key",
+			setup: func(t *testing.T, m *mockRouteProcessor,
+				req *BuildOnionRequest) {
+
+				m.unmarshallRoute = mockRoute
+			},
+			checkResponse: func(t *testing.T, resp *BuildOnionResponse) {
+				require.Len(t, resp.OnionBlob, lnwire.OnionPacketSize)
+				require.Len(t, resp.SessionKey, 32)
+				require.Len(t, resp.HopPubkeys, len(mockRoute.Hops))
+			},
+		},
+		{
+			name: "success with provided session key",
+			setup: func(t *testing.T, m *mockRouteProcessor,
+				req *BuildOnionRequest) {
+
+				m.unmarshallRoute = mockRoute
+				req.SessionKey = validSessionKeyBytes
+			},
+			checkResponse: func(t *testing.T, resp *BuildOnionResponse) {
+				require.Len(t, resp.OnionBlob, lnwire.OnionPacketSize)
+				require.Equal(t, validSessionKeyBytes, resp.SessionKey)
```

Comment:
nit: could also check the length of returned pubkeys like above, equality would be even better